feat: add nodejs to system install

dirien opened the pull request:
Signed-off-by: Engin Diri <engin.diri@mail.schwarz>

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

This PR adds node to the system install fn of arkade.

You can install from the stable release or choose a nightly build from the nightly channel

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [ ] I have raised an issue to propose this change, which has been given a label of `design/approved` by a maintainer ([required](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md))
Releates to #654 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

### Latest from release channel

```bash
❯ docker container run --rm -v $(pwd):/workspace -ti golang:1.16 bash
root@b2e70d12efcf:/go# cd /workspace/
root@b2e70d12efcf:/workspace# go build .
root@b2e70d12efcf:/workspace# ./arkade system install node
Installing Node.js to /usr/local/lib/nodejs
Installing version: v17.9.0 for: x64
Downloading from: https://nodejs.org/download/release/v17.9.0/node-v17.9.0-linux-x64.tar.gz
40.97 MiB / 40.97 MiB [--------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
Downloaded to: /tmp/node-v17.9.0-linux-x64.tar.gz
Unpacking binaries to: /tmp/node1248639963
2022/04/10 19:36:46 extracted tarball into /tmp/node1248639963: 4641 files, 1171 dirs (1.29905914s)
Copying binaries to: /usr/local/lib/nodejs
root@b2e70d12efcf:/workspace# node --version
v17.9.0
````
### From nightly channel

```bash
❯ docker container run --rm -v $(pwd):/workspace -ti golang:1.16 bash
root@08890b3376e3:/go# cd /workspace/
root@08890b3376e3:/workspace# go build .
root@08890b3376e3:/workspace# ./arkade system install node --version v18.0.0-nightly20220410a0461255c0  --channel nightly
Installing Node.js to /usr/local/lib/nodejs
Installing version: v18.0.0-nightly20220410a0461255c0 for: x64
Downloading from: https://nodejs.org/download/nightly/v18.0.0-nightly20220410a0461255c0/node-v18.0.0-nightly20220410a0461255c0-linux-x64.tar.gz
41.35 MiB / 41.35 MiB [--------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
Downloaded to: /tmp/node-v18.0.0-nightly20220410a0461255c0-linux-x64.tar.gz
Unpacking binaries to: /tmp/node4070689855
2022/04/10 19:40:01 extracted tarball into /tmp/node4070689855: 4639 files, 1172 dirs (1.545232732s)
Copying binaries to: /usr/local/lib/nodejs
root@08890b3376e3:/workspace# node --version
v18.0.0-nightly20220410a0461255c0
```

## Are you a GitHub Sponsor (Yes/No?)

<!--- Check at https://github.com/sponsors/alexellis         -->
<!--- Sponsors get priority because they support the project -->

- [ ] Yes
- [x] No

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Documentation

- [ ] I have updated the list of tools in README.md if (required) with `./arkade get -o markdown`
- [ ] I have updated the list of apps in README.md if (required) with `./arkade install --help`

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--- In case it is a new application -->
- [ ] I have tested this on arm, or have added code to prevent deployment

alexellis commented:
Thank you. Please can you try to match the style that I used in the new CNI plugins app or the Go app? I noticed Richard's PR had introduced lots of extracted variables and consts.

dirien commented:
@alexellis, Yes I try to align with your recent changes. 

Any particular code piece you saw? Would help me

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+	"strings"
+)
+
+func getLatestVersion(version, channel string) (*string, error) {
```
Given that this function is package scoped, it should be named uniquely, it can only get Node's version, so let's update it?

dirien replied:
done, renamed to `getLatestNodeVersion`

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+)
+
+func getLatestVersion(version, channel string) (*string, error) {
+	resp, err := http.Get(fmt.Sprintf("https://nodejs.org/download/%s/%s", channel, version))
```
As a nit, generally we have used `res` elsewhere, can you be consistent?

dirien replied:
done, renamed to `res`

alexellis replied:
I'm still seing it as `resp`?

dirien replied:
Done

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+	command := &cobra.Command{
+		Use:   "node",
+		Short: "Install Node.js",
+		Long:  `Install the Node.js JavaScript runtime ✨🐢🚀✨`,
```
The emojis are cute, but can you keep this consistent with the other apps?

The official homepage says: `"Node.js is a JavaScript runtime built on [Chrome's V8 JavaScript engine](https://v8.dev/).`"

dirien replied:
Just c&p the text from their github description! 😅 

I removed it

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+		Short: "Install Node.js",
+		Long:  `Install the Node.js JavaScript runtime ✨🐢🚀✨`,
+		Example: `arkade system install node
+  arkade system install node --version v17.9.0`,
```
Does `--version` also support `lts`? Given the layout of the website, I could see this as a common request.

dirien replied:
Nope, currently not. Should I add lts as alias too? Would mean some changes as i need to evaluate the json too.

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+		SilenceUsage: true,
+	}
+
+	command.Flags().StringP("version", "v", "latest", "The version for Node.js to install, either a specific version, 'latest' or ' latest-codename'")
```
What does `latest-codename` mean?

dirien replied:
this are the codename of the different versions:
![image](https://user-images.githubusercontent.com/38325136/162690417-6fed27e6-fcd0-4b42-8868-be76700aba90.png)

I thought, people may want to use them instead of a specific version or latest

What do youz think @alexellis?

alexellis replied:
latest-codename sounds like a value that you'd enter in, perhaps we could write:

`'latest' or a specific version`

alexellis replied:
Can you add a specific version into this list as an example?

alexellis replied:
Rather than "latest-codename", put `latest-CODENAME` so that it's obvious what you mean. I'd never heard of "codename" and have been installing Node.js for a number of years now.

dirien replied:
done

`The version for Node.js to install, either a specific version, 'latest' or 'latest-CODENAME' (eg: latest-gallium)`

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	command.Flags().StringP("version", "v", "latest", "The version for Node.js to install, either a specific version, 'latest' or ' latest-codename'")
+	command.Flags().StringP("path", "p", "/usr/local/lib/nodejs", "Installation path, where a go subfolder will be created")
```
The un-tar path for Nodejs on Linux is:

```
/usr/local/
```

Then the subfolders get created under that i.e. `./lib/...` `./bin/...`

```bash
pi@faasd-pi:~ $ wget https://nodejs.org/dist/v16.14.2/node-v16.14.2-linux-armv7l.tar.xz
--2022-04-11 08:44:57--  https://nodejs.org/dist/v16.14.2/node-v16.14.2-linux-armv7l.tar.xz
Resolving nodejs.org (nodejs.org)... 104.20.23.46, 104.20.22.46, 2606:4700:10::6814:172e, ...
Connecting to nodejs.org (nodejs.org)|104.20.23.46|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18275504 (17M) [application/x-xz]
Saving to: ‘node-v16.14.2-linux-armv7l.tar.xz’

node-v16.14.2-linux 100%[===================>]  17.43M  4.87MB/s    in 3.9s    

2022-04-11 08:45:01 (4.45 MB/s) - ‘node-v16.14.2-linux-armv7l.tar.xz’ saved [18275504/18275504]

pi@faasd-pi:~ $ sudo tar -xf node-v16.14.2-linux-armv7l.tar.xz -C /usr/local/ --strip-components=1
pi@faasd-pi:~ $ node --version
v16.14.2
pi@faasd-pi:~ $ 
```

Please match that style. The archive.UntarP function should cover this use-case.

dirien replied:
The problem I had, and thats the reason I worked around with the Copy function is that 

archive.untar is not honoring the Folders inside the archive and you get something like this:
![image](https://user-images.githubusercontent.com/38325136/162693881-4ca8a6ed-eef7-40e1-b0af-1862d0ad0cfa.png)

The issue with untar_nested is that you get a subfolder
![image](https://user-images.githubusercontent.com/38325136/162693974-4a08f681-fba1-4db3-8e7a-400ecf2af705.png)

I need a way to copy the content of the archive while honoring the structure and keeping the symlinks.

Maybe I am completly wrong here @alexellis. Help is appreciated.

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+
+	command.Flags().StringP("version", "v", "latest", "The version for Node.js to install, either a specific version, 'latest' or ' latest-codename'")
+	command.Flags().StringP("path", "p", "/usr/local/lib/nodejs", "Installation path, where a go subfolder will be created")
+	command.Flags().Bool("symlink", true, "Create symlinks in /usr/local/bin")
```
Not required, when the above is done.

dirien replied:
done

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+
+		fmt.Printf("Copying binaries to: %s\n", installPath)
+		nodeDir := fmt.Sprintf("%s/%s", tempUnpackPath, fmt.Sprintf("node-%s-linux-%s", version, dlArch))
+		err = cp.Copy(nodeDir, installPath)
```
You could probably use the tar methods provided rather than copying the files. Let me know how that goes, it'll be considerably simpler, but the archive method may not support the concept of stripping the prefix.

dirien replied:
done

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,138 @@
+		}
+		symlink, _ := cmd.Flags().GetBool("symlink")
+		if symlink {
+			os.Symlink(installPath+"/bin/node", "/usr/local/bin/node")
```
Not required when placed in the usual location.

dirien replied:
done!

alexellis commented on pkg/archive/untar_nested.go:
```diff
@@ -106,6 +106,10 @@ func untarNested(r io.Reader, dir string) (err error) {
 				return err
 			}
 			madeDir[abs] = true
+		case os.ModeSymlink != 0:
```
Did you run into a use-case that failed, and required this addition?

Is it in the upstream code?

https://cs.opensource.google/go/x/build/+/ef7e9155:internal/untar/untar.go;l=27

dirien replied:
The nodejs binary contains Symlinks for `npm` and `npx` which did not worked, when used the existing untar_nested.

alexellis replied:
Thanks

alexellis replied:
I ran into this with the github-actions runner too, which also bundles npm. :+1:

alexellis reviewed (changes requested):
Adding a few review comments.

dirien commented:
Hi @alexellis, removed the symlink stuff and added your feedback.

To only issue I have is, that I need to copy the content of the archive to the `/usr/local` folder.

Maybe you have an idea

alexellis commented:
> To only issue I have is, that I need to copy the content of the archive to the /usr/local folder.
> Maybe you have an idea

We have a couple of Untar methods to try out, what does each do vs. what we need?

Do we need a 3rd Untar method?

alexellis commented on cmd/system/node.go:
```diff
@@ -0,0 +1,125 @@
+
+		fmt.Printf("Copying binaries to: %s\n", installPath)
+		nodeDir := fmt.Sprintf("%s/%s", tempUnpackPath, fmt.Sprintf("node-%s-linux-%s", version, dlArch))
+		err = cp.Copy(nodeDir, installPath)
```
We may as well inline this:

```
if err : = cp.Copy(nodeDir, installPath) ; err !=nil {
    return err
}

dirien replied:
done!

alexellis reviewed (commented):
Do you think we can get this turned around this week?

alexellis reviewed (approved):
LGTM. Thank you very much for this.

alexellis commented:
Danke sehr.

Do you have a mode added already for quiet output?
